In [11]:
import pandas as pd 
import numpy as np 

import warnings
warnings.filterwarnings("ignore")

In [12]:
df = pd.read_csv('../data/Churn_Modelling_Cleaned.csv')
df.head()

CreditScore  Gender       Age  Tenure   Balance  NumOfProducts  HasCrCard  \
0        0.538       0  0.324324     0.2  0.000000              1          1   
1        0.516       0  0.310811     0.1  0.334031              1          0   
2        0.304       0  0.324324     0.8  0.636357              3          1   
3        0.698       0  0.283784     0.1  0.000000              2          0   
4        1.000       0  0.337838     0.2  0.500246              1          1   

   IsActiveMember  EstimatedSalary  Exited  Geography_Germany  Geography_Spain  
0               1         0.506735       1              False            False  
1               1         0.562709       0              False             True  
2               0         0.569654       1              False            False  
3               0         0.469120       0              False            False  
4               1         0.395400       0              False             True

### Splitting the training and testing data

In [13]:
X = df.drop('Exited', axis='columns')
y = df['Exited']

In [14]:
from sklearn.model_selection import train_test_split 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=5)

In [15]:
X_train.shape

(8000, 11)

In [16]:
X_test.shape

(2000, 11)

In [17]:
len(X_test.columns)

11

In [18]:
import tensorflow as tf 
from tensorflow import keras 
from sklearn.metrics import classification_report, confusion_matrix

In [19]:
def ANN(X_train, y_train, X_test, y_test, loss, weights):
    model = keras.Sequential([
        keras.layers.Dense(20, input_shape=(11,), activation='relu'), # "more hidden layers can be added" 
        keras.layers.Dense(1, activation='sigmoid')
    ])

    model.compile(optimizer='adam', 
              loss=loss, 
              metrics=['accuracy']) 
    
    if weights == -1:
        model.fit(X_train, y_train, epochs=100)

    else: 
        model.fit(X_train, y_train, epochs=100, class_weights=weights)

    print("Model evaluation: \n", model.evaluate(X_test, y_test))

    y_preds = model.predict(X_test)
    y_preds = np.round(y_preds)

    print("Classification report: \n", classification_report(y_test, y_preds))

    return y_preds

In [20]:
y_preds = ANN(X_train, y_train, X_test, y_test, 'binary_crossentropy', -1)

Epoch 1/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7561 - loss: 0.5667
Epoch 2/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7912 - loss: 0.4858
Epoch 3/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8010 - loss: 0.4565
Epoch 4/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8022 - loss: 0.4537
Epoch 5/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8101 - loss: 0.4405
Epoch 6/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8134 - loss: 0.4364
Epoch 7/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8169 - loss: 0.4299
Epoch 8/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8200 - loss: 0.4264
Epoch 9/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8214 - loss: 0.4178
Epoch 10/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8149 - loss: 0.4193
Epoch 11/100
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8139 - loss: 0.4211
Epoch 12/100
250/250 ━━━━━━━━━━━━━━━━━━━━

Low recall for class 1 (44%) – The model fails to identify a significant portion of the minority class instances, leading to many false negatives.  

Class imbalance – There are 1595 instances of class 0 vs. only 405 of class 1, likely causing the model to favor class 0.  

Precision-recall tradeoff – The model prioritizes precision (74%) over recall (44%) for class 1, making it conservative in predicting positive cases.